### PR TITLE
[AIRFLOW-6352] security - ui - add login timeout

### PR DESCRIPTION
**Pull request by tooptoop4**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-6352\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-6352
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-6352\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by potiuk** [comment]
Same here @tooptoop4 -> I love the small fixes you provide but static checks are failing for them (and could be prevented easily by pre-commit).

**Comment by tooptoop4** [comment]
travis unrelated?

**Comment by potiuk** [comment]
This change introduces a new behaviour (logout after 60 minutes).  While it is good for  security reasons (obviously) the UI of Airflow has a little bit different use patterns/characteristics than typical user-facing apps. 

It's mostly internal use, with very small number of users, it's already behind a VPN and I guess often witht some kind of client certificates being verified by web seervers. I can imagine in those cases prolonged session persistency might be important feature for users using Apache Airflow. In many cases UI of Airflow can be used in a fashion similar to "operational dashboard" rather than the typical case of "login/do something/logout".

Since we have no auto-refresh yet, using Airflow as dashboard with 60 minutes logout session would not be super user-friendly. 

I'd love to hear what others think about it, but I believe at the very least UPDATING.md should mention that new behaviour if we agree this is a good thing to introduce 60 minutes (or another period) timeout.

**Comment by RosterIn** [comment]
I think it can become handy.
@potiuk Security isn't always about the "outside" world. It's also for inner one. Companies may use more than one instance of airflow. For example I have airflow install unique for HR data - this is extremely sensitive. I would welcome such feature for this instance but I wouldn't welcome it for my other airflow instances.

So I think this shouldn't be enforced for everyone.
This can be optional feature that users can set by themselves from `airflow.cfg`
I also recommend not to hard code 60 min. Let users decide what they want. It can be in the same setting:

`force_log_out_after = 60` if set this will timeout the session and also use the number to determine the time

**Comment by potiuk** [comment]
Agree @RosterIn with security. Internal security should not be neglected. 

It's just that security is never an on/off switch and "let's apply all the possible security practices" is not always best choice.

There are often multiple layers of security in different places so this logout might not be needed (for example when you have individual client certificates individually issued to your users and verified in proxy standing in front of Airflow.). This is often case in corporate environments (such as Airflow often is installed at) but almost unheard of in public-facing websites.

There is always a delicate balance "convenience vs. security" and sometimes enforcing some "best practices" for security with some inconveniences built in gives the opposite result. People tend to bypass security inconveniences by introducing even more insecure workarounds. For example in this case, I can very easily imagine a data engineer wanting a dashboard installing "auto-refresh" browser plugin to refresh the airflow dashboard every 20 minutes. Been there, done that. Such plugins are often vectors of attack on their own.

So yeah I agree with force_log_out_after conf value. I think having a separate conf entry for that is much better choice and gives freedom to admins to set their policy rules as they find best for their users.

**Review by potiuk — CHANGES_REQUESTED** [review]
So to sum-up @tooptoop4 -> this is a good idea but we should make it configurable :).

**Comment by tooptoop4** [comment]
@potiuk make sense, I've now guarded with config

**Review by turbaszek — APPROVED** [review]

**Review by potiuk — APPROVED** [review]
Great!

File: airflow/config_templates/default_airflow.cfg [hunk 1]
```diff
@@ -370,10 +370,13 @@ default_wrap = False
 # on webserver startup
 update_fab_perms = True
 
+# Minutes of non-activity before logged out from UI
+# 0 means never get forcibly logged out
```
Comment (mik-laj) [inline airflow/config_templates/default_airflow.cfg:374]: Can you add some documentation about this option?   If there is no information about this feature in the documentation, very few people will be able to use it.

Comment (potiuk) [reply airflow/config_templates/default_airflow.cfg:374]: @mik-laj - > where do you think the documentation should be added ? Maybe you can provide some pointers? Most of the config options are documented in comments of the default_airlfow.cfg. Any other specific place you think this documentation should be added (and where all the previous options are documented?)

Comment (mik-laj) [reply airflow/config_templates/default_airflow.cfg:374]: Related discussion: https://github.com/apache/airflow/pull/6952#discussion_r362102402

I think, we can add short description in 'security.rst' file or a new file in `howto` directory.

**Comment by tooptoop4** [comment]
@mik-laj any idea how this works? these lines are missing from app.py in 1.10.9
import datetime
import flask
import flask_login

**Comment by NBardelot** [comment]
@tooptoop4 sadly it doesn't work as-is (see https://issues.apache.org/jira/browse/AIRFLOW-6865).

**Comment by dephusluke** [comment]
I'm trying also to understand the origin of this bug. The imports appear in the file so how could they be missed?

I see also there is discussion about it in the commit https://github.com/apache/airflow/commit/0721d62bf1b23a471d7fe97ccf108db9f92403ef

@tooptoop4 have you figured it out?

**Comment by renaldrozario** [comment]
This issue is still open with version 1.10.12.. can someone advise when this will be fixed?

**Comment by mik-laj** [comment]
@renaldrozario  We are now focusing on the development of Airflow 2.0 and we are limiting the new features that are released in Airflow 1.10. Airflow 2.0 will be released at the end of the year.

**Comment by mik-laj** [comment]
Is this information helpful to you?

**Pull request merged**